src/weak_memory.rs

```diff
@@ -0,0 +1,271 @@
+    fn get_store_buffer(&self, range: AllocRange) -> Ref<'_, StoreBuffer> {
+        Ref::map(self.store_buffer.borrow(), |range_map| {
+            let (.., store_buffer) = range_map.iter(range.start, range.size).next().unwrap();
+            store_buffer
+        })
+    }
+
+    fn get_store_buffer_mut(&self, range: AllocRange) -> RefMut<'_, StoreBuffer> {
+        RefMut::map(self.store_buffer.borrow_mut(), |range_map| {
+            let (.., store_buffer) = range_map.iter_mut(range.start, range.size).next().unwrap();
+            store_buffer
+        })
+    }
+
+    /// Reads from the last store in modification order
+    pub fn read_from_last_store<'tcx>(&self, range: AllocRange, global: &GlobalState) {
+        let store_buffer = self.get_store_buffer(range);
+        let store_elem = store_buffer.buffer.back();
+        if let Some(store_elem) = store_elem {
+            let (index, clocks) = global.current_thread_state();
+            store_elem.load_impl(index, &clocks);
+        }
+    }
+
+    pub fn buffered_read<'tcx>(
+        &self,
+        range: AllocRange,
+        global: &GlobalState,
+        is_seqcst: bool,
+        rng: &mut (impl rand::Rng + ?Sized),
+        validate: impl FnOnce() -> InterpResult<'tcx>,
+    ) -> InterpResult<'tcx, Option<ScalarMaybeUninit<Tag>>> {
+        // Having a live borrow to store_buffer while calling validate_atomic_load is fine
+        // because the race detector doesn't touch store_buffer
+        let store_buffer = self.get_store_buffer(range);
+
+        let store_elem = {
+            // The `clocks` we got here must be dropped before calling validate_atomic_load
+            // as the race detector will update it
+            let (.., clocks) = global.current_thread_state();
+            // Load from a valid entry in the store buffer
+            store_buffer.fetch_store(is_seqcst, &clocks, &mut *rng)
+        };
+
+        // Unlike in write_scalar_atomic, thread clock updates have to be done
+        // after we've picked a store element from the store buffer, as presented
+        // in ATOMIC LOAD rule of the paper. This is because fetch_store
+        // require access to ThreadClockSet.clock, which is updated by the race detector
+        validate()?;
+
+        let loaded = store_elem.map(|store_elem| {
+            let (index, clocks) = global.current_thread_state();
+            store_elem.load_impl(index, &clocks)
+        });
+        Ok(loaded)
+    }
+
+    pub fn buffered_write<'tcx>(
+        &mut self,
+        val: ScalarMaybeUninit<Tag>,
+        range: AllocRange,
+        global: &GlobalState,
+        is_seqcst: bool,
+    ) -> InterpResult<'tcx> {
+        let (index, clocks) = global.current_thread_state();
+
+        let mut store_buffer = self.get_store_buffer_mut(range);
+        store_buffer.store_impl(val, index, &clocks.clock, is_seqcst);
+        Ok(())
+    }
+}
+
+const STORE_BUFFER_LIMIT: usize = 128;
```

Comment:
How was this value picked?

Reply:
From the paper
> *Bounding of store and load buffers* To prevent unbounded mem-
ory overhead, we must bound the size of store buffers so that the
oldest element of a full buffer is evicted when a new store element
is pushed. This restricts the stores that loads can read from, so the
buffer size trades memory overhead for observable behaviours. For
our evaluation we used a buffer size of 128 to allow a relatively
wide range of stores to be available to load operations. Load buffers
need not be bounded. This is because at most one load element per
thread is required for any store element: the oldest load has the
smallest epoch, so if a later load blocks a thread, so will the oldest

Reply:
Okay, so this is a completeness tradeoff?
Please briefly mention that (and that we follow the paper with this value) in a comment.